File: android/src/main/java/com/unact/yandexmapkit/PolygonMapObjectController.java
package com.unact.yandexmapkit;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.PolygonMapObject;

import java.lang.ref.WeakReference;
import java.util.Map;

public class PolygonMapObjectController extends MapObjectController implements MapObjectTapListener {
  public final PolygonMapObject polygon;
  private boolean consumeTapEvents = false;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public PolygonMapObjectController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    PolygonMapObject polygon = parent.addPolygon(
      Utils.polygonFromJson((Map<String, Object>) params.get("polygon"))
    );

    this.polygon = polygon;
    this.id = (String) params.get("id");
    this.controller = controller;

    polygon.setUserData(this.id);
    polygon.addTapListener(this);
    update(params);
  }

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public void update(Map<String, Object> params) {
    polygon.setGeometry(Utils.polygonFromJson((Map<String, Object>) params.get("polygon")));
    polygon.setGeodesic((Boolean) params.get("isGeodesic"));
    polygon.setZIndex(((Double) params.get("zIndex")).floatValue());
    polygon.setVisible((Boolean) params.get("isVisible"));
    polygon.setStrokeWidth(((Double) params.get("strokeWidth")).floatValue());
    polygon.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    polygon.setFillColor(((Number) params.get("fillColor")).intValue());

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    polygon.getParent().remove(polygon);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}

File: ios/Classes/PolygonMapObjectController.swift
import YandexMapsMobile

class PolygonMapObjectController: NSObject, MapObjectController, YMKMapObjectTapListener {
  public let polygon: YMKPolygonMapObject
  private var consumeTapEvents: Bool = false
  public weak var controller: YandexMapController?
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let polygon = parent.addPolygon(with: Utils.polygonFromJson(params["polygon"] as! [String: Any]))

    self.polygon = polygon
    self.id = params["id"] as! String
    self.controller = controller

    super.init()

    polygon.userData = self.id
    polygon.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    polygon.geometry = Utils.polygonFromJson(params["polygon"] as! [String: Any])
    polygon.isGeodesic = (params["isGeodesic"] as! NSNumber).boolValue
    polygon.zIndex = (params["zIndex"] as! NSNumber).floatValue
    polygon.isVisible = (params["isVisible"] as! NSNumber).boolValue
    polygon.strokeColor = Utils.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value)
    polygon.strokeWidth = (params["strokeWidth"] as! NSNumber).floatValue
    polygon.fillColor = Utils.uiColor(fromInt: (params["fillColor"] as! NSNumber).int64Value)

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    polygon.parent.remove(with: polygon)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller!.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}
